src/mon/PGMonitor.cc

```diff
@@ -1132,47 +1139,58 @@ void PGMonitor::map_pg_creates()
       &acting,
       &acting_primary);
 
-    if (s->acting_primary != -1) {
-      pg_map.creating_pgs_by_osd[s->acting_primary].erase(pgid);
-      if (pg_map.creating_pgs_by_osd[s->acting_primary].size() == 0)
-        pg_map.creating_pgs_by_osd.erase(s->acting_primary);
-    }
-    s->up = up;
-    s->up_primary = up_primary;
-    s->acting = acting;
-    s->acting_primary = acting_primary;
-    pg_map.stat_pg_add(pgid, *s, true);
+    if (up != s->up ||
+	up_primary != s->up_primary ||
+	acting !=  s->acting ||
+	acting_primary != s->acting_primary) {
+      dout(20) << __func__ << "  " << pgid << " "
+	       << " primary: " << s->acting_primary << " -> " << acting_primary
+	       << " acting: " << s->acting << " -> " << acting
+	       << dendl;
```

Comment:
We should print out the up[_primary] as well, since it can lead us into this branch?
Also the spacing doesn't seem to match on these lines.